File: express/src/api/routes/places.ts
import { Router, Request, Response, NextFunction } from "express";
import { crudPlace } from "../../models/crud";
import {
    z,
    zodObjectId,
    PlaceSearchSchema,
    PlaceSortableFields,
    PlacePost,
    PlacePostSchema,
    PlacePut,
    PlacePutSchema,
    PlaceSearchSwagger,
    PlacesPaginatedSchema,
    PlaceReadSchema,
} from "../../models/schemas";
import { validateBody, filter, Authenticated } from "../middlewares";
import { swaggerRegistery } from "../openapi";

export const placeRouter = Router();

// Get Places Endpoint
async function getPlaces(req: Request, resp: Response, next: NextFunction) {
    // All places are public
    const query = req.filterQuery!;
    resp.status(200).json(await crudPlace.fetch(query));
}

placeRouter.get("/", filter(PlaceSearchSchema, PlaceSortableFields), getPlaces);

swaggerRegistery.registerPath({
    method: "get",
    path: "/places/",
    request: {
        query: PlaceSearchSwagger,
    },
    responses: {
        200: {
            description: "Search and Filter places",
            content: {
                "application/json": {
                    schema: PlacesPaginatedSchema,
                },
            },
        },
    },
    tags: ["Place"],
    summary: "Search and Retrieve places",
});

// Post New Places
async function createPlace(req: Request, resp: Response, next: NextFunction) {
    // Use safeCreate to avoid a user posting a place for another user
    const parsed = req.parsed as PlacePost;
    const currentUser = req.currentUser!;
    const newPlace = await crudPlace.safeCreate(currentUser, parsed);
    resp.status(200).json(newPlace);
}

placeRouter.post(
    "/",
    Authenticated,
    validateBody(PlacePostSchema),
    createPlace
);

swaggerRegistery.registerPath({
    method: "post",
    path: "/places/",
    request: {
        body: {
            content: {
                "multipart/form-data": {
                    schema: PlacePostSchema,
                },
            },
            description: "Place creation",
            required: true,
        },
    },
    responses: {
        200: {
            description: "Place creation",
            content: {
                "application/json": {
                    schema: PlaceReadSchema,
                },
            },
        },
    },
    tags: ["Place"],
    summary: "Place Creation",
    security: [
        {
            BearerAuth: [],
        },
    ],
});

// Get a place by ID
async function getPlace(req: Request, res: Response, next: NextFunction) {
    // All places are public
    const place = await crudPlace.get(req.params.placeId);
    res.status(200).json(place);
}

placeRouter.get("/:placeId", getPlace);

swaggerRegistery.registerPath({
    method: "get",
    path: "/places/{placeId}",
    request: {
        params: z.object({
            placeId: zodObjectId().openapi({
                example: "507f1f77bcf86cd799439011",
                description: "Place Id",
            }),
        }),
    },
    responses: {
        200: {
            description: "Place data",
            content: {
                "application/json": {
                    schema: PlaceReadSchema,
                },
            },
        },
    },
    tags: ["Place"],
    summary: "Search and Retrieve place by id",
});

// Edit Places
async function editPlace(req: Request, res: Response, next: NextFunction) {
    const parsed = req.parsed as PlacePut;
    const currentUser = req.currentUser!;
    const updatedPlace = await crudPlace.safeUpdateById(
        currentUser,
        req.params.placeId,
        parsed
    );
    res.status(200).json(updatedPlace);
}

placeRouter.put(
    "/:placeId",
    Authenticated,
    validateBody(PlacePutSchema),
    editPlace
);

swaggerRegistery.registerPath({
    method: "put",
    path: "/places/{placeId}",
    request: {
        params: z.object({
            placeId: zodObjectId().openapi({
                example: "507f1f77bcf86cd799439011",
                description: "Place id",
            }),
        }),
        body: {
            content: {
                "application/json": {
                    schema: PlacePutSchema,
                },
            },
            description: "Place updated data",
            required: true,
        },
    },
    responses: {
        200: {
            description: "Place data",
            content: {
                "application/json": {
                    schema: PlaceReadSchema,
                },
            },
        },
    },
    tags: ["Place"],
    summary: "Update places",
    security: [
        {
            BearerAuth: [],
        },
    ],
});

// Delete Places
async function deletePlace(req: Request, res: Response, next: NextFunction) {
    const currentUser = req.currentUser!;
    await crudPlace.safeDelete(currentUser, req.params.placeId);
    res.status(200).json({
        message: `Deleted place ${req.params.placeId}`,
    });
}

placeRouter.delete("/:placeId", Authenticated, deletePlace);

swaggerRegistery.registerPath({
    method: "delete",
    path: "/places/{placeId}",
    request: {
        params: z.object({
            placeId: zodObjectId().openapi({
                example: "507f1f77bcf86cd799439011",
                description: "Place Id",
            }),
        }),
    },
    responses: {
        200: {
            description: "Deletion confirmation message",
            content: {
                "application/json": {
                    schema: z.object({
                        message: z.string().openapi({
                            example: "Deleted place 507f1f77bcf86cd799439011",
                        }),
                    }),
                },
            },
        },
    },
    tags: ["Place"],
    summary: "Delete place by id",
    security: [
        {
            BearerAuth: [],
        },
    ],
});

File: backend-gin/internal/api/routes/places.go
package routes

import (
	"backend/internal/api/middlewares"
	"backend/internal/lib/gin_"
	"backend/internal/models/cruds"
	"backend/internal/models/schemas"
	"fmt"
	"net/http"

	"github.com/gin-gonic/gin"
)

// @Summary      Search and Filter places
// @Tags         Place
// @Accept       application/json
// @Produce      json
// @Security     OAuth2Password[admin]
// @Param        params query schemas.PlaceFilters true "GET parameters"
// @Success      200  {object}  schemas.PlacesPaginated
// @Router       /api/places/ [get]
func getPlaces(c *gin.Context) {
	searchQuery, _ := gin_.GetSearchQuery(c)
	cp := cruds.GetCRUDSPlace()
	data, err := cp.Paginate(searchQuery, &cruds.PlaceOptions{Process: true})
	if err != nil {
		gin_.AbortWithStatusJSON(c, err)
		return
	}
	c.JSON(http.StatusOK, data)
}

// @Summary      Search and Retrieve places
// @Tags         Place
// @Accept       application/json
// @Produce      json
// @Security     OAuth2Password[admin]
// @Param        params body schemas.PlaceFilters true "POST parameters"
// @Success      200  {object}  schemas.PlacesPaginated
// @Router       /api/places/search [post]
func searchPlaces(c *gin.Context) {
	searchQuery, _ := gin_.GetSearchQuery(c)
	cp := cruds.GetCRUDSPlace()
	data, err := cp.Paginate(searchQuery, &cruds.PlaceOptions{Process: true})
	if err != nil {
		gin_.AbortWithStatusJSON(c, err)
		return
	}
	c.JSON(http.StatusOK, data)
}

// @Summary      Place creation
// @Tags         Place
// @Accept       mpfd
// @Produce      json
// @Security     OAuth2Password[admin]
// @Param        params formData schemas.PlacePost true "POST parameters"
// @Param        image  formData file   true "Place Image (JPEG)"
// @Success      200  {object}  schemas.PlaceRead
// @Router       /api/places [post]
func createPlace(c *gin.Context) {
	currentUser := c.MustGet("currentUser").(schemas.UserRead)
	body, _ := gin_.GetBody[schemas.PlacePost](c)

	cp := cruds.GetCRUDSPlace()
	place, err := cp.UserPost(currentUser, body, &cruds.PlaceOptions{Process: true})
	if err != nil {
		gin_.AbortWithStatusJSON(c, err)
		return
	}
	c.JSON(http.StatusOK, place)
}

// @Summary      Search and Retrieve place by id
// @Tags         Place
// @Accept       application/json
// @Produce      json
// @Security     OAuth2Password[admin]
// @Param        placeId path string true "Place ID"
// @Success      200  {object}  schemas.PlaceRead
// @Router       /api/places/{placeId} [get]
func getPlace(c *gin.Context) {
	placeId, err := gin_.ExtractId(c, "placeId")
	if err != nil {
		gin_.AbortWithStatusJSON(c, err)
		return
	}

	cp := cruds.GetCRUDSPlace()
	place, err := cp.GetPartial(placeId, &cruds.PlaceOptions{Process: true})
	if err != nil {
		gin_.AbortWithStatusJSON(c, err)
		return
	}
	c.JSON(http.StatusOK, place)
}

// @Summary      Update places
// @Tags         Place
// @Accept       application/json
// @Produce      json
// @Security     OAuth2Password[admin]
// @Param        placeId path string true "Place ID"
// @Param        params body schemas.PlacePut true "PUT parameters"
// @Success      200  {object}  schemas.PlaceRead
// @Router       /api/places/{placeId} [put]
func updatePlace(c *gin.Context) {
	currentUser := c.MustGet("currentUser").(schemas.UserRead)
	body, _ := gin_.GetBody[schemas.PlacePut](c)

	placeId, err := gin_.ExtractId(c, "placeId")
	if err != nil {
		gin_.AbortWithStatusJSON(c, err)
		return
	}

	cp := cruds.GetCRUDSPlace()
	place, err := cp.UserPut(currentUser, placeId, body, &cruds.PlaceOptions{Process: true})
	if err != nil {
		gin_.AbortWithStatusJSON(c, err)
		return
	}
	c.JSON(http.StatusOK, place)
}

type PlaceDeleteResponse struct {
	Message string `json:"message" example:"Deleted place 507f1f77bcf86cd799439011"`
}

// @Summary      Delete place by id
// @Tags         Place
// @Accept       application/json
// @Produce      json
// @Security     OAuth2Password[admin]
// @Param        placeId path string true "Place ID"
// @Success      200  {object}  PlaceDeleteResponse
// @Router       /api/places/{placeId} [delete]
func deletePlace(c *gin.Context) {
	currentUser := c.MustGet("currentUser").(schemas.UserRead)

	placeId, err := gin_.ExtractId(c, "placeId")
	if err != nil {
		gin_.AbortWithStatusJSON(c, err)
		return
	}

	cp := cruds.GetCRUDSPlace()
	err = cp.UserDelete(currentUser, placeId)
	if err != nil {
		gin_.AbortWithStatusJSON(c, err)
		return
	}

	message := fmt.Sprintf("Deleted place %s", placeId)
	c.JSON(http.StatusOK, gin.H{"message": message})
}

func RegisterPlaces(r *gin.Engine) {
	router := r.Group("/api/places")
	router.GET(
		"/",
		middlewares.Authenticated,
		gin_.QueryFilters[schemas.PlaceSearch],
		getPlaces,
	)
	router.POST(
		"/search",
		middlewares.Authenticated,
		gin_.BodyFilters[schemas.PlaceSearch],
		searchPlaces,
	)
	router.POST(
		"",
		middlewares.Authenticated,
		gin_.BodyValidator[schemas.PlacePost],
		createPlace,
	)
	router.GET(
		"/:placeId",
		middlewares.Authenticated,
		gin_.QueryValidator[schemas.PlaceGet],
		getPlace,
	)
	router.PUT(
		"/:placeId",
		middlewares.Authenticated,
		gin_.BodyValidator[schemas.PlacePut],
		updatePlace,
	)
	router.DELETE("/:placeId", middlewares.Authenticated, deletePlace)
}
